### PR TITLE
Data loss protect resending

### PR DESCRIPTION
**Pull request by halseth**
This PR makes the latest channel reestablishment message being stored as part of the channel close summary, such that it can be sent even when a channel ahs been closed.

This makes it possible for nodes to recover from the case where they have been offline, lost state, and comes back online, attempting to reseync the channel.

Note: This PR contains a DB migration. To ensure deserialization code is not changed in the future, and breaking old migrations, a versioned copy of the current deserialization logic is added to the file `legacy_serialization.go`. Backwards compatible addition of the new field turned out to be complex because of the presence of optional fields (see the now removed commit https://github.com/lightningnetwork/lnd/commit/14648b236fa5f203a76e876d75026735f2356d1f).

TODO:
- [x] Unit test for `FetchClosedChannelForID`

### CODE REVIEW

File: channeldb/channel.go [hunk 1]
```diff
@@ -2203,6 +2224,32 @@ func deserializeCloseChannelSummary(r io.Reader) (*ChannelCloseSummary, error) {
 		}
 	}
 
+	// Check if we have a channel sync message to read.
+	var hasChanSyncMsg bool
+	err = ReadElements(r, &hasChanSyncMsg)
+	if err == io.EOF {
```
Comment (halseth) [inline channeldb/channel.go]: Note: we can add write the `false` marker here as part of the migration to avoid this `EOF` check.

Comment (halseth) [reply channeldb/channel.go]: Currently the addition of the `LastChanSyncMsg` field is kept in its own commit to show how new fields can easily be added after the presence flags are introduced in the serialization.

File: channeldb/channel.go [hunk 2]
```diff
@@ -2203,6 +2224,32 @@ func deserializeCloseChannelSummary(r io.Reader) (*ChannelCloseSummary, error) {
...
+		return c, nil
+	} else if err != nil {
+		return nil, err
+	}
+
+	// If a chan sync message is present, read it.
+	if hasChanSyncMsg {
+		// We must pass in reference to a lnwire.Message for the codec
+		// to support it.
+		var msg lnwire.Message
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply channeldb/channel.go]: i'm guessing that `**lnwire.ChannelReestablish` and `*lnwire.Message` are distinct concrete types, and the type switch doesn't try to determine if the suffix `*lnwire.ChannelReestablish` implements the other suffix `lnwire.Message`

[the remark this answers is not in the document]
Comment (halseth) [reply channeldb/channel.go]: Yes, I think that's right. I think the type switch is checking if the passed `interface{}` is of concrete type `*lnwire.Message`, not if it is satisfying the interface. TIL you cannot do that: https://play.golang.org/p/SmL37Hobvvs

File: channeldb/db.go [hunk 3]
```diff
@@ -616,6 +617,55 @@ func (d *DB) FetchClosedChannel(chanID *wire.OutPoint) (*ChannelCloseSummary, er
 	return chanSummary, nil
 }
 
+// FetchClosedChannelForID queries for a channel close summary using the
+// channel ID of the channel in question.
+func (d *DB) FetchClosedChannelForID(cid lnwire.ChannelID) (
```
Comment (Roasbeef) [inline channeldb/db.go]: This as is scales with the number of closed channels, which will increase for all time. Instead, we can use a different algorithm, which will be constant-time ish, as in we know the hard upper bound

  1. Fetch the entire set of channel point candidates for a channel ID via `GenPossibleOutPoints`
  2. Run through each of these, and attempt to read the key from the `closedChannelBucket`

This should terminate in 1, and sometimes two iterations. Only once people start doing this like splicing or funding coin joins will this start to swell. Even then, it should be well below 1000 or so iterations.

File: channeldb/channel.go [hunk 4]
```diff
@@ -3,6 +3,7 @@ package channeldb
 import (
 	"bytes"
 	"encoding/binary"
+	"errors"
```
Comment (Roasbeef) [inline channeldb/channel.go]: Lingering TODO in the commit.

File: channeldb/channel.go (hunk 1 above)
Comment (Roasbeef) [reply channeldb/channel.go]: Why are we mixing the two approaches? We should either use EOF everywhere (so no migration) or the bool based approach.

Comment (Roasbeef) [reply channeldb/channel.go]: Ah scratch that, I see the usage now

File: contractcourt/chain_watcher.go [hunk 5]
```diff
@@ -488,6 +489,19 @@ func (c *chainWatcher) toSelfAmount(tx *wire.MsgTx) btcutil.Amount {
...
 
+// chanSyncMsg fetches the latest channel sync message for the channel.
+func (c *chainWatcher) chanSyncMsg() (*lnwire.ChannelReestablish, error) {
+	lnChannel, err := lnwallet.NewLightningChannel(
+		nil, nil, c.cfg.chanState,
+	)
+	if err != nil {
+		return nil, err
+	}
+	defer lnChannel.Stop()
```
Comment (Roasbeef) [inline contractcourt/chain_watcher.go]: Bit of a code smell that we need to do this (make a set of goroutines) just to grab this field. Instead we can [efactor this method slightly](https://github.com/lightningnetwork/lnd/blob/5803da382a4d16c155cd5396d585c0340736ecad/lnwallet/channel.go#L3449:29) to make a (possible private version) that accepts the channel state and commitments heights and use that in place. Recall the issues we had in the funding manager that caused us to just pass around the channel state everywhere vs the full state machine.

File: contractcourt/chain_watcher.go [hunk 6]
```diff
@@ -590,6 +613,15 @@ func (c *chainWatcher) dispatchLocalForceClose(
 		closeSummary.TimeLockedBalance += htlcValue
 	}
 
+	// Attempt to add a channel sync message to the close summary.
```
Comment (Roasbeef) [inline contractcourt/chain_watcher.go]: What about remote force close?

File: contractcourt/chain_watcher.go [hunk 7]
```diff
@@ -749,6 +781,15 @@ func (c *chainWatcher) dispatchContractBreach(spendEvent *chainntnfs.SpendDetail
...
+	chanSync, err := c.chanSyncMsg()
+	if err != nil {
+		log.Errorf("ChannelPoint(%v): unable to create channel sync "+
+			"message: %v", c.cfg.chanState.FundingOutpoint, err)
+	} else {
+		closeSummary.LastChanSyncMsg = chanSync
+	}
+
```
Comment (Roasbeef) [inline contractcourt/chain_watcher.go]: Unrelated to this PR, but shouldn't the `BRAR` be the one that's making this DB state here? Re the whole reliable hand off thing.

File: lnd_test.go [hunk 8]
```diff
@@ -276,6 +276,58 @@ func closeChannelAndAssert(ctx context.Context, t *harnessTest,
 	return closingTxid
 }
 
+// waitForChannelPendingForceClose waits for the node to report that the
```
Comment (Roasbeef) [inline lnd_test.go]: This commit can be deleted when rebasing to master as it was merged on its own.

File: peer.go [hunk 9]
```diff
@@ -868,7 +868,33 @@ func newChanMsgStream(p *peer, cid lnwire.ChannelID) *msgStream {
...
+				switch {
+
+				// If we failed to find the link in question,
+				// and the message received was a channel sync
+				// message, then this might be a peer trying to
+				// resync closed channel. In this case we'll
+				// try to resend our last channel sync message,
+				// such that the peer can recover funds from
+				// the closed channel.
+				case err != nil && isChanSyncMsg:
```
Comment (Roasbeef) [inline peer.go]: 👍

File: contractcourt/chain_watcher.go [hunk 10]
```diff
@@ -403,16 +404,41 @@ func (c *chainWatcher) closeObserver(spendNtfn *chainntnfs.SpendEvent) {
...
-				log.Errorf("Unable to retrieve commitment "+
-					"point for channel(%v) with lost "+
-					"state: %v",
-					c.cfg.chanState.FundingOutpoint, err)
-				return
+			// can sweep our funds. If we cannot find the commit
+			// point, there's not much we can do other than wait
+			// for us to retrieve it. We will attempt to retrieve
+			// it from the peer each time we connect to it.
+			// TODO(halseth): actively initiate re-connection to
```
Comment (Roasbeef) [inline contractcourt/chain_watcher.go]: This can eventually be hooked up into the channel life cycle management stuff. So to request a notification once a new node connects. Actually we already have this for use in the funding manager, so perhaps it can be used here?

File: lnd_test.go [hunk 11]
```diff
@@ -7198,6 +7198,85 @@ func testDataLossProtection(net *lntest.NetworkHarness, t *harnessTest) {
 
 	assertNodeNumChannels(t, ctxb, dave, 0)
 	assertNodeNumChannels(t, ctxb, carol, 0)
+
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply lnd_test.go]: Least blocking route would be a comment that delineates this new testing scenario.

File: contractcourt/chain_watcher.go (hunk 7 above)
Comment (cfromknecht) [reply contractcourt/chain_watcher.go]: the brar writes the retributions to disk, then acks back. after that the chain watcher knows it's safe to mark the channel as pending closed

File: contractcourt/chain_watcher.go (hunk 10 above)
Comment (cfromknecht) [reply contractcourt/chain_watcher.go]: yeah i think we can use `NotifyWhenOnline` to just try if they ever reconnect

File: channeldb/db.go (hunk 3 above)
Comment (Roasbeef) [reply channeldb/db.go]: Actually scratch that, I properly see the seek!

File: channeldb/channel.go [hunk 12]
```diff
@@ -2094,7 +2094,12 @@ func serializeChannelCloseSummary(w io.Writer, cs *ChannelCloseSummary) error {
 	// If this is a close channel summary created before the addition of
 	// the new fields, then we can exit here.
 	if cs.RemoteCurrentRevocation == nil {
-		return nil
+		return WriteElements(w, false)
```
[the remark this answers is not in the document]
Comment (halseth) [reply channeldb/channel.go]: The problem with optional fields is that it only works for the last field. If there are more than one optional field, then you won't know what the data present represents. This way we attempt to make it future proof.

By adding booleans to indicate presence, we can make sure old code will read the new fields (they won't attempt to read any extra data they don't understand) and that new code will read the old format (booleans will indicate whether fields are present, so it can know which fields are not present from the old format).

File: channeldb/channel.go [hunk 13]
```diff
@@ -2105,14 +2110,28 @@ func serializeChannelCloseSummary(w io.Writer, cs *ChannelCloseSummary) error {
...
-	// We'll write this field last, as it's possible for a channel to be
-	// closed before we learn of the next unrevoked revocation point for
-	// the remote party.
+	// The RemoteNextRevocation field is optional, as it's possible for a
+	// channel to be closed before we learn of the next unrevoked
+	// revocation point for the remote party. Write a boolen indicating
+	// whether this field is present or not.
 	if cs.RemoteNextRevocation == nil {
-		return nil
+		if err := WriteElements(w, false); err != nil {
```
[the remark this answers is not in the document]
Comment (halseth) [reply channeldb/channel.go]: Won't work for the case where `cs.RemoteNextRevocation == nil` and `chanSyncMsg != nil`, as the deserialization code will attempt to read the `chanSyncMsg` as `RemoteNextRevocation`.

File: channeldb/migrations.go [hunk 14]
```diff
@@ -573,3 +573,40 @@ func migratePruneEdgeUpdateIndex(tx *bolt.Tx) error {
 
 	return nil
 }
+
+// migrateOptionalChannelCloseSummaryFields migrates the serialized format of
```
[the remark this answers is not in the document]
Comment (halseth) [reply channeldb/migrations.go]: I first attempted to do this without a migration, but it turned out to be complex: https://github.com/lightningnetwork/lnd/commit/14648b236fa5f203a76e876d75026735f2356d1f

File: channeldb/channel.go (hunk 1 above)
Comment (halseth) [reply channeldb/channel.go]: This is how you would add new fields in the future without a migration.

File: channeldb/channel.go (hunk 4 above)
Comment (halseth) [reply channeldb/channel.go]: I think it makes sense to NOT do it in the migration (the only difference would be that we could remove the EOF check below), so will remove the TODO from the commit message if no protests.

File: contractcourt/chain_watcher.go (hunk 7 above)
Comment (halseth) [reply contractcourt/chain_watcher.go]: We could let the `BRAR` have that responsibility, as we could get rid of ACK logic.

File: contractcourt/chain_watcher.go (hunk 5 above)
Comment (halseth) [reply contractcourt/chain_watcher.go]: Fixed!

File: contractcourt/chain_watcher.go (hunk 6 above)
Comment (halseth) [reply contractcourt/chain_watcher.go]: Good catch! Fixed.

File: lnd_test.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (halseth) [reply lnd_test.go]: Improved the comments.

Agreed that some of the tests would benefit from being split up. I think this might cause our integration tests to run even longer though, bc of the added overhead. Might be worth looking into in combination with running integration tests in parallel (perhaps using containers?).

File: lnd_test.go (hunk 8 above)
Comment (halseth) [reply lnd_test.go]: Deleted.

File: lnd_test.go [hunk 15]
```diff
@@ -6839,183 +6891,229 @@ func testDataLossProtection(net *lntest.NetworkHarness, t *harnessTest) {
...
-		ctxt, t, net, carol, dave,
-		lntest.OpenChannelParams{
-			Amt: chanAmt,
-		},
-	)
+	// We create a helper method that will create a slice of payment
+	// requests for us. The incrementing invoiceIndex will ensure they are
+	// unique.
+	invoiceIndex := 0
+	createPayReqs := func(node *lntest.HarnessNode,
```
[the remark this answers is not in the document]
Comment (halseth) [reply lnd_test.go]: Done!

File: contractcourt/chain_watcher.go (hunk 10 above)
Comment (halseth) [reply contractcourt/chain_watcher.go]: I actually like the current polling more, since we might need some time to process the `ChannelReestablish` after they reconnect. In that case the first lookup after they connect might fail, so we have to wait a little bit anyway. 

If we don't wait before retrying at that point, we risk the peer coming online send us the `ChannelReestablish` and disconnects, and we will stall here waiting for them to connect again.

**Comment by halseth** [comment]
Comments address, PTAL 🐶

File: contractcourt/chain_watcher.go (hunk 7 above)
Comment (cfromknecht) [reply contractcourt/chain_watcher.go]: the cleaner solution IMO is just to write resolutions in `CloseChannel`. then there is no need for a reliable handoff. the resolutions can be passed in memory to kick off the process, but if we crash the brar would look for all breached channels in the db and load resolutions from there

File: channeldb/channel.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (Roasbeef) [reply channeldb/channel.go]: Yep, realized this later that day. I think in the future, we should do a clean sweep to just move to a TLV format everywhere. You read it all, then at the end check the array/map for the keys you actually know of then parse that. With that, we gain a way to handle an arbitrary number of fields in the future in a uniform manner throughout the codebase.

File: channeldb/migrations.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (Roasbeef) [reply channeldb/migrations.go]: Ah ok, yeah that's pretty convoluted.

File: lnd_test.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (Roasbeef) [reply lnd_test.go]: 👍 

Yeah we'll need to white board out a better architecture as we add more backends, and also eventually starting other platforms other than mac OS.

File: channeldb/channel.go (hunk 4 above)
Comment (Roasbeef) [reply channeldb/channel.go]: SGTM

File: channeldb/db_test.go [hunk 16]
```diff
@@ -71,3 +74,76 @@ func TestWipe(t *testing.T) {
 			ErrNoClosedChannels, err)
 	}
 }
+
+// TestFetchClosedChannelForID tests that we are able to properly retrieve a
```
Comment (Roasbeef) [inline channeldb/db_test.go]: 👍

File: contractcourt/chain_watcher.go [hunk 17]
```diff
@@ -402,16 +403,39 @@ func (c *chainWatcher) closeObserver(spendNtfn *chainntnfs.SpendEvent) {
...
-			if err != nil {
+			// can sweep our funds. If we cannot find the commit
+			// point, there's not much we can do other than wait
+			// for us to retrieve it. We will attempt to retrieve
+			// it from the peer each time we connect to it.
+			// TODO(halseth): actively initiate re-connection to
+			// the peer?
+			var commitPoint *btcec.PublicKey
+			backoff := time.Second
+			maxBackoff := time.Hour
```
Comment (Roasbeef) [inline contractcourt/chain_watcher.go]: This seems a bit long of a max back off, perhaps something in the 10 of minutes instead?

File: contractcourt/chain_watcher.go (hunk 6 above)
Comment (Roasbeef) [reply contractcourt/chain_watcher.go]: Don't see it fixed in the latest diff: https://github.com/lightningnetwork/lnd/blob/4124c6db8159bf79327181029f613fc9d404f4a1/contractcourt/chain_watcher.go#L654

File: lnwallet/channel.go [hunk 18]
```diff
@@ -1023,11 +1023,10 @@ type commitmentChain struct {
...
 }
 
-// newCommitmentChain creates a new commitment chain from an initial height.
-func newCommitmentChain(initialHeight uint64) *commitmentChain {
+// newCommitmentChain creates a new commitment chain.
+func newCommitmentChain() *commitmentChain {
 	return &commitmentChain{
-		commitments:    list.New(),
-		startingHeight: initialHeight,
+		commitments: list.New(),
```
Comment (Roasbeef) [inline lnwallet/channel.go]: Where do we set the new starting height of a commitment chain? Don't see why this change is necessary atm. There're other spots where we depend on the height being set properly.

File: lnwallet/channel.go [hunk 19]
```diff
@@ -5113,6 +5112,15 @@ func NewUnilateralCloseSummary(chanState *channeldb.OpenChannel, signer Signer,
 		LocalChanConfig:         chanState.LocalChanCfg,
 	}
 
+	// Attempt to add a channel sync message to the close summary.
+	chanSync, err := ChanSyncMsg(chanState)
```
Comment (Roasbeef) [inline lnwallet/channel.go]: Ah ok I see this now re remote close. We could also shuffle the other spots we populate this information into this package. Don't consider it a blocker though.

**Comment by Roasbeef** [comment]
Final lingering comments are re if we should have a lower backoff, and also if a minor refactor is necessary or not given.

File: lnwallet/channel.go [hunk 20]
```diff
@@ -3446,27 +3445,27 @@ func (lc *LightningChannel) ProcessChanSyncMsg(
 //      it.
 //   3. We didn't get the last RevokeAndAck message they sent, so they'll
 //      re-send it.
-func (lc *LightningChannel) ChanSyncMsg() (*lnwire.ChannelReestablish, error) {
+func ChanSyncMsg(c *channeldb.OpenChannel) (*lnwire.ChannelReestablish, error) {
```
Comment (cfromknecht) [inline lnwallet/channel.go]: should this have a mutex around it? esp now that this is called in more places than just the switch

also seems like something that should be implemented in channeldb as a method of `OpenChannel`, since it is otherwise unrelated to anything in `lnwallet`. would be nice to keep the `.ChanSyncMsg()` notation, and not use a package level function

File: channeldb/migrations.go [hunk 21]
```diff
@@ -573,3 +573,40 @@ func migratePruneEdgeUpdateIndex(tx *bolt.Tx) error {
...
+// ChannelCloseSummary to a format where optional fields' presence is indicated
+// with boolean markers.
+func migrateOptionalChannelCloseSummaryFields(tx *bolt.Tx) error {
```
Comment (cfromknecht) [inline channeldb/migrations.go]: would be nice to have a unit test for this migration, asserting it behaves properly in all four cases

File: channeldb/channel.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (halseth) [reply channeldb/channel.go]: ooooh, yeah that would be nice!

File: lnwallet/channel.go (hunk 20 above)
Comment (halseth) [reply lnwallet/channel.go]: Good idea, added mutex.

The reason I didn't make this a method on `OpenChannel` is that it calls `ComputeCommitmentPoint`, which is a method from `lnwallet`, leading to an import cycle. We could however move this method to a new package (`lnutil`? 😛) if we want this, but I felt it was not worth it.

Also moving it to `channeldb` felt wrong 🤔

File: contractcourt/chain_watcher.go [hunk 22]
```diff
@@ -403,16 +404,41 @@ func (c *chainWatcher) closeObserver(spendNtfn *chainntnfs.SpendEvent) {
...
+			for {
```
[the remark this answers is not in the document]
Comment (halseth) [reply contractcourt/chain_watcher.go]: kewl, sticking with the fixup for now

File: lnwallet/channel.go (hunk 18 above)
Comment (halseth) [reply lnwallet/channel.go]: We never read it, we just look at the height of the commitment in the chain.

Agreed that the change is not directly relevant to this PR, I just noticed when transforming the `ChanSync` method to not being dependent on the chain heights.

Removed and made a separate PR: https://github.com/lightningnetwork/lnd/pull/2206

File: channeldb/legacy_serialization.go [hunk 23]
```diff
@@ -0,0 +1,53 @@
+package channeldb
+
+import "io"
+
+// deserializeCloseChannelSummaryV6 reads the v6 database format for
+// ChannelCloseSummary.
+//
+// NOTE: deprecated, only for migration.
+func deserializeCloseChannelSummaryV6(r io.Reader) (*ChannelCloseSummary, error) {
```
Comment (halseth) [inline channeldb/legacy_serialization.go]: This was after an idea by @joostjager. Lmk what you think :)

File: channeldb/migrations.go (hunk 21 above)
Comment (halseth) [reply channeldb/migrations.go]: Migration added! (only three cases though?)

**Comment by halseth** [comment]
Comments addressed and commits squashed. PTAL.
